### ISSUE TITLE
Grad_norm is nan and Loss is 0: 'loss': 0.0, 'grad_norm': nan

### ISSUE THREAD

**Issue by qm-intel** [LLaVA-VL/LLaVA-NeXT#435]
Hi, 

I am using two datasets for training a `Qwen/Qwen2-7B` model, which should also support multi-image inputs. 
Datasets are:

1. LLaVA 

   - LLaVA-Pretrain (`blip_laion_cc_sbu_558k.json`)
   - LLaVA-InstructionTuning (`llava_v1_5_mix665k.json`)

2. Custom dataset
   - pertaining data
   - instruction-tuning data
This dataset includes multi-image inputs, which the format looks like the following (I am not sure if this format is correct):
```
{
    "image": [
      "images/im_2_0.jpg",
      "images/im_2_1.jpg"
    ],
    "conversations": [
      {
        "from": "human",
        "value": "What is depicted in the images?\n<image>\n<image>"
      },.....
```
The following is the order of training: 

**1-1)** First, I pre-trained on the `LLaVA-Pretrain` dataset, and the following is the `pretrain_clip.sh` script:

```
#export OMP_NUM_THREADS=8
#export NCCL_IB_DISABLE=0
#export NCCL_IB_GID_INDEX=3
##export NCCL_SOCKET_IFNAME=eth0
#export NCCL_DEBUG=INFO

LLM_VERSION="Qwen/Qwen2-7B" #"Qwen/Qwen2-7B-Instruct"


LLM_VERSION_CLEAN="${LLM_VERSION//\//_}"
VISION_MODEL_VERSION="openai/clip-vit-large-patch14-336"
VISION_MODEL_VERSION_CLEAN="${VISION_MODEL_VERSION//\//_}"


############### Pretrain ################

PROMPT_VERSION=plain

BASE_RUN_NAME="llava-${VISION_MODEL_VERSION_CLEAN}-${LLM_VERSION_CLEAN}-mlp2x_gelu-pretrain_blip558k_plain"



echo "BASE_RUN_NAME: ${BASE_RUN_NAME}"

##LlaVA Dataset
PATH_TO_PRETRAIN_DATA="~/datasets/mllm/LLaVA-Pretrain/blip_laion_cc_sbu_558k.json"
PATH_TO_IMAGE_DATA="~/datasets/mllm/LLaVA-Pretrain/images"  


PATH_TO_OUTPUT_DIRECTORY="./checkpoints/projectors/${BASE_RUN_NAME}"


NUM_GPUS=1
NNODES=1
RANK=0
ADDR=127.0.0.1
PORT=29500

#export CUDA_VISIBLE_DEVICES=0
#deepspeed --master_port 29510 --include  localhost:0,1 \
ACCELERATE_CPU_AFFINITY=1 torchrun --nproc_per_node="${NUM_GPUS}" --nnodes="${NNODES}" --node_rank="${RANK}" --master_addr="${ADDR}" --master_port="${PORT}" \
    llava/train/train_mem.py \
    --deepspeed scripts/zero2.json \
    --model_name_or_path ${LLM_VERSION} \
    --version ${PROMPT_VERSION} \
    --data_path $PATH_TO_PRETRAIN_DATA \
    --image_folder $PATH_TO_IMAGE_DATA \
    --vision_tower ${VISION_MODEL_VERSION} \
    --mm_tunable_parts="mm_mlp_adapter" \
    --mm_vision_select_layer -2 \
    --mm_projector_type mlp2x_gelu \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --bf16 True \
    --output_dir ${PATH_TO_OUTPUT_DIRECTORY} \
    --num_train_epochs 1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 32 \
    --evaluation_strategy "no" \
    --save_strategy "no" \
    --save_steps 50000 \
    --learning_rate 1e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length 8192 \
    --gradient_checkpointing True \
    --dataloader_num_workers 16 \
    --lazy_preprocess True \
    --report_to none \
    --run_name $BASE_RUN_NAME \
    --attn_implementation sdpa
```

**1-2)** Finetune using lora on `LLaVA-InstructionTuning`:

```
#export OMP_NUM_THREADS=8
#export NCCL_IB_DISABLE=0
#export NCCL_IB_GID_INDEX=3
#export NCCL_SOCKET_IFNAME=eth0
#export NCCL_DEBUG=INFO

LLM_VERSION="Qwen/Qwen2-7B"

LLM_VERSION_CLEAN="${LLM_VERSION//\//_}"
VISION_MODEL_VERSION="openai/clip-vit-large-patch14-336"
VISION_MODEL_VERSION_CLEAN="${VISION_MODEL_VERSION//\//_}"

############### Pretrain ################

PROMPT_VERSION="qwen_1_5"

BASE_RUN_NAME="llava-${VISION_MODEL_VERSION_CLEAN}-${LLM_VERSION_CLEAN}-mlp2x_gelu-pretrain_blip558k_plain"

echo "BASE_RUN_NAME: ${BASE_RUN_NAME}"

PATH_TO_DATA="~/datasets/mllm/LLaVA-InstructionTuning/llava_v1_5_mix665k.json"
PATH_TO_IMAGE="~/datasets/mllm/LLaVA-InstructionTuning"


PATH_TO_MM_ADAPTER="./checkpoints/projectors/${BASE_RUN_NAME}/mm_projector.bin"

MID_RUN_NAME="llava_lora_finetune_qwen2_7b_mix665k"

export CUDA_VISIBLE_DEVICES=0

NUM_GPUS=1
NNODES=1
RANK=0
ADDR=127.0.0.1
PORT=29500

#--mm_tunable_parts="mm_vision_tower,mm_mlp_adapter,mm_language_model" \
#    --mm_vision_tower_lr=2e-6 \
#    --image_aspect_ratio anyres \
#    --image_grid_pinpoints "[(336, 672), (672, 336), (672, 672), (1008, 336), (336, 1008)]" \
#    --mm_patch_merge_type spatial_unpad \
#    --save_steps 3000 \
#    --learning_rate 1e-5 \
#    --torch_compile True \
#    --torch_compile_backend "inductor" \
#    --lora_enable True --lora_r 128 --lora_alpha 256 --mm_projector_lr 2e-5 \

ACCELERATE_CPU_AFFINITY=1 torchrun --nproc_per_node="${NUM_GPUS}" --nnodes="${NNODES}" --node_rank="${RANK}" --master_addr="${ADDR}" --master_port="${PORT}" \
    llava/train/train_mem.py \
    --lora_enable True --lora_r 128 --lora_alpha 256 --mm_projector_lr 2e-5 \
    --deepspeed scripts/zero2.json \
    --model_name_or_path ${LLM_VERSION} \
    --version ${PROMPT_VERSION} \
    --data_path $PATH_TO_DATA \
    --image_folder $PATH_TO_IMAGE \
    --pretrain_mm_mlp_adapter $PATH_TO_MM_ADAPTER \
    --mm_tunable_parts="mm_mlp_adapter" \
    --vision_tower ${VISION_MODEL_VERSION} \
    --mm_projector_type mlp2x_gelu \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --group_by_modality_length True \
    --image_aspect_ratio pad \
    --bf16 True \
    --run_name $MID_RUN_NAME \
    --output_dir "./checkpoints/${MID_RUN_NAME}" \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 64 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 2e-4 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length 8192 \
    --gradient_checkpointing True \
    --dataloader_num_workers 16 \
    --lazy_preprocess True \
    --report_to none \
    --dataloader_drop_last True \
    --attn_implementation sdpa
```
For lora fine-tuning, I have added the following command before `LlavaTrainer()`:

```
# Re-enable LoRA parameters if LoRA is enabled
    if training_args.lora_enable:
        for name, param in model.named_parameters():
            if "lora_" in name:
                param.requires_grad = True
```
**1-3) Merge Lora weights**
Once fine-tuning on llava dataset is done, I merge the lora weights:L

```
python ./scripts/archived/merge_lora_weights.py --model-path checkpoints/llava_lora_finetune_qwen2_7b_mix665k --model-base Qwen/Qwen2-7B --save-model-path  checkpoints/llava-Qwen_Qwen2-7B-merged
```

I added the following to the `builder.py` after line #80:

```
elif "qwen" in model_name.lower():  #+
                import pdb
                from llava.model.language_model.llava_qwen import LlavaQwenConfig
                lora_cfg_pretrained = LlavaQwenConfig.from_pretrained(model_path)

                tokenizer = AutoTokenizer.from_pretrained(model_base, use_fast=False)
                model = LlavaQwenForCausalLM.from_pretrained(model_base, low_cpu_mem_usage=True, config=lora_cfg_pretrained, 
                                  attn_implementation=attn_implementation, **kwargs)
```
Now when I am trying to pre-train and fine-tune the second dataset by setting `LLM_VERSION="checkpoints/llava-Qwen_Qwen2-7B-merged"`, I am getting for example this is for 500 samples only

```
Building extension module fused_adam...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module fused_adam...
Time to load fused_adam op: 0.0200653076171875 seconds
  0%|                                                                                                                                                                                                | 0/1 [00:00<?, ?it/s]/home/user/anaconda3/envs/llava-next/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:745: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. In version 2.5 we will raise an exception if use_reentrant is not passed. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.
  return fn(*args, **kwargs)
/home/user/anaconda3/envs/llava-next/lib/python3.10/site-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
{'loss': 0.0, 'grad_norm': nan, 'learning_rate': 0.001, 'epoch': 0.51}                                                                                                                                                     
{'train_runtime': 83.4181, 'train_samples_per_second': 5.994, 'train_steps_per_second': 0.012, 'train_loss': 0.0, 'epoch': 0.51}                                                                                           
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [01:23<00:00, 83.42s/it]
Rank 0:  Only save projectors: True
Rank 0:  Model saved to ./checkpoints/projectors/....
```

**How to resolve this issue? what is causing this error?**

I also followed some suggestion here: 
https://github.com/deepspeedai/DeepSpeed/issues/5242

but it still having the same issue after merging the lora weights and trying to pretrain and fine-tune on the second dataset.
